### PR TITLE
#0: Use TensorLayout in Tensor

### PR DESCRIPTION
**Pull request by sminakov-tt**
### Problem description
We're migrating TTNN infrastructure to use new TensorLayout everywhere, including Tensor

### What's changed
Introduced `TensorSpec`, which represents `logical_shape` + `tensor_layout`
Used `TensorSpec` in Tensor instead of `(shape, dtype, layout, tile)`
Refactored Tensor constructors
Use `set_tensor_spec` instead of a bunch of different setters, removed the need of manually setting `metadata_populated = true`

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/11960140767)
- [ ] Blackhole Post commit (if applicable)
- [x] [Model regression CI testing passes](https://github.com/tenstorrent/tt-metal/actions/runs/11956334891)
- [x] [Device performance regression CI testing passes](https://github.com/tenstorrent/tt-metal/actions/runs/11956337771)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/tensor/tensor.hpp [hunk 1]
```diff
@@ -198,20 +201,18 @@ struct Tensor {
...
-    inline const tt::tt_metal::LegacyShape &legacy_shape() const { return this->tensor_attributes->shape.value; };
-    inline const ttnn::Shape &shape() const { return this->tensor_attributes->shape; };
-    inline const DataType &dtype() const { return this->tensor_attributes->dtype; };
-    inline const Layout &layout() const { return this->tensor_attributes->layout; };
-    inline const Tile &tile() const { return this->tensor_attributes->tile; };
+    inline tt::tt_metal::LegacyShape legacy_shape() const { return this->tensor_attributes->tensor_spec.compute_shape().value; };
+    inline ttnn::Shape shape() const { return this->tensor_attributes->tensor_spec.compute_shape(); };
+    inline DataType dtype() const { return this->tensor_attributes->tensor_spec.tensor_layout().get_data_type(); };
+    inline Layout layout() const { return this->tensor_attributes->tensor_spec.tensor_layout().get_layout(); };
+    inline Tile tile() const { return this->tensor_attributes->tensor_spec.tensor_layout().get_page_config().get_tile(); };
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/tensor.hpp:208]: lets discuss this. ideally we should not calculate these all the time, they are not changing dynamically

Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/tensor.hpp:208]: The only calculated one out of theses getters is `shape()`. I'm planning to look into how often each `compute_*` method is called and cache some of them.

File: ttnn/cpp/ttnn/operation.hpp [hunk 2]
```diff
@@ -413,11 +413,8 @@ auto default_create_output_tensors(
     const auto& device = input_tensors.at(0).device();
     const auto& output_specs = operation.compute_output_specs(input_tensors);
     output_tensors.reserve(output_specs.size());
-    for (const auto& [output_shape, output_layout] : output_specs) {
-        output_tensors.emplace_back(create_device_tensor(
-            output_shape,
-            output_layout,
-            device));
+    for (const auto& output_spec : output_specs) {
+        output_tensors.emplace_back(create_device_tensor(output_spec, device));
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operation.hpp:417]: unexpectedly nice

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 3]
```diff
@@ -696,17 +696,14 @@ std::pair<ttnn::Tensor, std::optional<ttnn::Tensor>> prepare_conv_weights_biases
...
+    auto target_shape = ttnn::Shape(std::array<uint32_t,4>{1, 1, weight_matrix_height, out_channels},
         std::array<std::array<uint32_t, 2>, 4>{
             std::array<uint32_t, 2>{0, 0},
             std::array<uint32_t, 2>{0, 0},
             std::array<uint32_t, 2>{0, weight_matrix_height_padding},
             std::array<uint32_t, 2>{0, out_channel_padding}
-    }));
-
+    });
+    weight_tensor_ = ttnn::reshape(weight_tensor_, target_shape);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:706]: @mywoodstock please take a look

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp [hunk 4]
```diff
@@ -920,8 +920,8 @@ Matmul create_matmul_struct(
     bool broadcast_batch =
         parameters.bcast_batch.value_or(get_broadcast_batch(input_tensor_a, input_tensor_b, parameters.program_config));
     TT_FATAL(!(has_user_grid && has_program_config), "Cannot use both user core grid/coordinates and a program config");
-    const auto& in0_tile = input_tensor_a.get_tile();
-    const auto& in1_tile = input_tensor_b.get_tile();
+    auto in0_tile = input_tensor_a.get_tile();
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp:923]: what do you think about this change?

File: ttnn/cpp/ttnn/run_operation_inl.hpp [hunk 5]
```diff
@@ -232,10 +232,7 @@ void launch_op(
                         insert_buffer_and_shape_for_device(target_device, *local_tensor, *output_tensor);
                         int num_workers_completed = (output_tensor->tensor_attributes->num_workers_completed)++;
                         if (not num_workers_completed) {
-                            output_tensor->tensor_attributes->shape = local_tensor->tensor_attributes->shape;
-                            output_tensor->tensor_attributes->dtype = local_tensor->tensor_attributes->dtype;
-                            output_tensor->tensor_attributes->layout = local_tensor->tensor_attributes->layout;
-                            output_tensor->tensor_attributes->tile = local_tensor->tensor_attributes->tile;
+                            output_tensor->set_tensor_spec(local_tensor->tensor_spec());
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/run_operation_inl.hpp:235]: not a good public method, would like to discuss this.

File: ttnn/cpp/ttnn/tensor/layout/page_config.cpp [hunk 6]
```diff
@@ -64,13 +64,13 @@ bool PageConfig::is_row_major() const {
...
 
-std::optional<Tile> PageConfig::get_tile() const
+Tile PageConfig::get_tile() const
 {
     if(std::holds_alternative<TilePageConfig>(config_)) {
         return std::get<TilePageConfig>(config_).get_tile();
     }
 
-    return std::nullopt;
+    return Tile();
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/page_config.cpp:73]: would like to discuss why

File: ttnn/cpp/ttnn/tensor/layout/page_config.cpp [hunk 7]
```diff
@@ -18,7 +18,7 @@ size_t element_size_bytes(DataType dtype) {
         case DataType::UINT8: return sizeof(uint8_t);
         case DataType::BFLOAT8_B:
         case DataType::BFLOAT4_B:
-            TT_THROW("element_size_bytes() should not be used for BFLOAT8_B and BFLOAT4_B types becaues of how they are packed");
+            return sizeof(float);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/page_config.cpp:21]: would like to discuss why

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp [hunk 8]
```diff
@@ -36,6 +36,12 @@ Alignment legacyShapeToAlignment(const ttnn::Shape& shape) {
         values[i] = legacy_padded_shape[i] * values[i + 1];
     }
 
+    for (auto& value : values) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp:39]: will this conflict with @TT-BrianLiu 's , change?

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.hpp [hunk 9]
```diff
@@ -51,6 +51,15 @@ class TensorLayout {
     //  H is all dimensions except W multiplied and aligned to tile and shard height
     Size compute_physical_shape(const ttnn::SimpleShape& shape) const;
 
+    void set_memory_config(MemoryConfig memory_config) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/tensor_layout.hpp:54]: wouldn't want it to be mutable like this. lets discuss

File: ttnn/cpp/ttnn/tensor/tensor_spec.hpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ttnn/tensor/types.hpp"
+#include "ttnn/tensor/layout/tensor_layout.hpp"
+
+namespace tt::tt_metal {
+
+class TensorSpec final {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/tensor_spec.hpp:12]: tbd, discuss next steps with caching some of the values

File: ttnn/cpp/ttnn/run_operation_inl.hpp (hunk 5 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/run_operation_inl.hpp:235]: I agree. It's used because the shape of the tensor isn't initially know. I believe we can fix this after all ops are migrated to `compute_output_specs`.
I do believe it beats setting 5 fields by hand though.

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp (hunk 8 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp:39]: I'm unsure. I would say this PR in general is likely to conflict with Brian's PR. Let's merge his PR first and I'll resolve the conflicts

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.hpp (hunk 9 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/layout/tensor_layout.hpp:54]: I agree. I needed to create exactly the same TensorLayout, but with a different memory config. I didn't figure out a good way to do it. I can create a method `TensorLayout with_memory_config(MemoryConfig) const;`, what do you think?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 3 above)
Comment (pavlejosipovic) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:706]: Can we make reshape run on device?

File: ttnn/cpp/ttnn/run_operation_inl.hpp (hunk 5 above)
Comment (bbradelTT) [reply ttnn/cpp/ttnn/run_operation_inl.hpp:235]: If you're planning to get rid of the method, there are ways of trying to keep the usage to a a minimum (private, protected, friend, deprecated).

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/tensor/layout/page_config.cpp (hunk 6 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/layout/page_config.cpp]: This code changed, added a comment and a deprecation warning related to this

File: ttnn/cpp/ttnn/tensor/layout/page_config.hpp [hunk 11]
```diff
@@ -20,11 +20,25 @@ namespace tt::tt_metal {
...
+
+    const Tile& get_tile() const;
+
+    static constexpr auto attribute_names = std::forward_as_tuple("tile");
+    const auto attribute_values() const {
+        return std::forward_as_tuple(tile_);
+    }
+
+private:
+    // This is currently needed for compatibility reasons.
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/page_config.hpp:39]: to discuss :)

File: ttnn/cpp/ttnn/tensor/tensor.hpp [hunk 12]
```diff
@@ -184,11 +181,12 @@ struct Tensor {
...
-    const Layout &get_layout() const;
-    const Tile &get_tile() const;
+    tt::tt_metal::LegacyShape get_legacy_shape() const;
+    ttnn::Shape get_shape() const;
+    DataType get_dtype() const;
+    Layout get_layout() const;
+    Tile get_tile() const;
+    const TensorSpec& get_tensor_spec() const;
 
     ttnn::SimpleShape get_logical_shape() const;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/tensor.hpp:191]: should be const ref

File: ttnn/cpp/ttnn/tensor/tensor.hpp [hunk 13]
```diff
@@ -184,11 +181,12 @@ struct Tensor {
...
-    const tt::tt_metal::LegacyShape &get_legacy_shape() const;
-    const ttnn::Shape &get_shape() const;
-    const DataType &get_dtype() const;
-    const Layout &get_layout() const;
-    const Tile &get_tile() const;
+    tt::tt_metal::LegacyShape get_legacy_shape() const;
+    ttnn::Shape get_shape() const;
+    DataType get_dtype() const;
+    Layout get_layout() const;
+    Tile get_tile() const;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/tensor.hpp:188]: should be const ref

Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/tensor.hpp:188]: must be removed and accessed from TensorSpec or TensorLayout. worst case optional.

File: ttnn/cpp/ttnn/tensor/layout/page_config.cpp [hunk 14]
```diff
@@ -37,10 +38,13 @@ PageConfig::PageConfig(Layout layout)
 
 PageConfig::PageConfig(Layout layout, const std::optional<Tile>& tile) {
     if(layout == Layout::ROW_MAJOR) {
-        config_ = RowMajorPageConfig();
+        if (tile.has_value()) {
+            tt::log_warning("Specifying tile shape for a row major layout is deprecated, and will be removed soon");
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/page_config.cpp:42]: Fire a ticket, attach here.

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/tensor/layout/page_config.cpp:42]: CC @yugaoTT

File: ttnn/cpp/ttnn/tensor/layout/page_config.hpp [hunk 15]
```diff
@@ -20,11 +20,25 @@ namespace tt::tt_metal {
...
 public:
+    RowMajorPageConfig(const Tile& tile = Tile());
+
     Alignment create_default_alignment(DataType dtype, const MemoryConfig& memory_config) const;
     void validate_alignment(const Alignment& alignment, DataType dtype, const MemoryConfig& memory_config) const;
 
     Size get_page_shape(const Size& physical_size, DataType dtype, const MemoryConfig& memory_config) const;
     size_t get_page_size_bytes(const Size& page_size, DataType dtype) const;
+
+    const Tile& get_tile() const;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/page_config.hpp:31]: deprecated and a comment

**Review by ayerofieiev-tt — APPROVED** [review]

**Comment by ayerofieiev-tt** [comment]
@sminakov-tt , please make sure to run 

- Model regression CI testing passes (if applicable)
- Device performance regression CI testing passes (if applicable)
- Nightly fast dispatch

Thank you!

File: ttnn/cpp/ttnn/operations/ccl/ccl_common.hpp [hunk 16]
```diff
@@ -412,17 +412,19 @@ class InterleavedRingAllGatherTensorSlicer : public LegacyCclTensorSlicer {
                     output_shape.begin() + slice_dim, output_shape.end() - 1, 1, std::multiplies<uint32_t>()) -
                 num_rows;
         } else {
-            this->num_cols = input_tensor.get_legacy_shape()[-1] / input_tensor.get_tile().get_width();
             auto input_shape = input_tensor.get_legacy_shape();
             auto output_shape = output_tensor.get_legacy_shape();
-            uint32_t num_output_cols = output_tensor.get_legacy_shape()[-1] / output_tensor.get_tile().get_width();
+            auto input_tile = input_tensor.tensor_spec().tile();
+            auto output_tile = output_tensor.tensor_spec().tile();
+            this->num_cols = input_shape[-1] / input_tensor.tensor_spec().tile().get_width();
```
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/ccl/ccl_common.hpp:419]:
```suggestion
            this->num_cols = input_shape[-1] / input_tile.get_width();
```
I think this was missed

**Review by SeanNijjar — APPROVED** [review]

File: tt_metal/tt_stl/reflection.hpp [hunk 17]
```diff
@@ -1041,7 +1041,8 @@ inline hash_t hash_object(const std::variant<Ts...>& variant) noexcept {
     if constexpr (DEBUG_HASH_OBJECT_FUNCTION) {
         fmt::print("Hashing std::variant: {}\n", variant);
     }
-    return std::visit([](const auto& value) { return hash_object(value); }, variant);
+    auto active_variant = variant.index();
+    return std::visit([&](const auto& value) { return hash_objects(active_variant, value); }, variant);
```
Comment (yan-zaretskiy) [inline tt_metal/tt_stl/reflection.hpp:1045]: Could you elaborate why this change was introduced?

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp (hunk 8 above)
Comment (TT-BrianLiu) [reply ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp:39]: I merged. Try rebasing? I don't think it's an issue

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp [hunk 18]
```diff
@@ -59,7 +65,9 @@ TensorLayout::TensorLayout(DataType dtype, const PageConfig& page_config, const
 }
 
 TensorLayout TensorLayout::fromLegacyPaddedShape(DataType dtype, const PageConfig& page_config, const MemoryConfig& memory_config, const ttnn::Shape& legacy_shape) {
-    return TensorLayout(dtype, page_config, memory_config, CMAKE_UNIQUE_NAMESPACE::legacyShapeToAlignment(legacy_shape));
+    auto result =  TensorLayout(dtype, page_config, memory_config);
+    result.alignment_ = CMAKE_UNIQUE_NAMESPACE::legacyShapeToAlignment(legacy_shape);
+    return result;
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp:70]: This will get default alignment and validate on it (I didn't really think through this logic if `legacyShapeToAlignment` is expected to be called later). Before it would just validate the passed in alignment from `legacyShapeToAlignment`. Is there a reason for this change?

File: ttnn/cpp/ttnn/tensor/tensor_spec.hpp [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    }
+    size_t compute_packed_buffer_size_bytes() const {
+        return tensor_layout_.compute_packed_buffer_size_bytes(logical_shape_);
+    }
+    size_t compute_page_size_bytes() const {
+        return tensor_layout_.compute_page_size_bytes(logical_shape_);
+    }
+    Size compute_physical_shape() const {
+        return tensor_layout_.compute_physical_shape(logical_shape_);
+    }
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/tensor/tensor_spec.hpp]: Should these functions be implemented at the tensor spec level if they all take in logical shape?

**Review by TT-BrianLiu — APPROVED** [review]

**Review by yugaoTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/repeat/repeat.cpp [hunk 20]
```diff
@@ -99,7 +99,9 @@ ttnn::Tensor RepeatOperation::invoke(
 
             auto padded_to_tiled_shape = ttnn::Shape(sliced_logical_shape.view(),
                                                      tiled_output.get_padded_shape().view());
-            tiled_output.set_shape(padded_to_tiled_shape);
+            std::cout << "repeat ???? called" << std::endl;
```
Comment (cfjchu) [inline ttnn/cpp/ttnn/operations/data_movement/repeat/repeat.cpp]: delete?

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 21]
```diff
@@ -696,16 +696,18 @@ std::pair<ttnn::Tensor, std::optional<ttnn::Tensor>> prepare_conv_weights_biases
     int32_t weight_matrix_height_padding = weight_tensor_.shape()[2] - weight_matrix_height;
     TT_FATAL(weight_matrix_height_padding >= 0," Matrix Height Padding can't be negative");
 
+    std::cout << "conv2d ???? called" << std::endl;
```
Comment (cfjchu) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp]: remove here too

File: ttnn/cpp/ttnn/tensor/layout/page_config.cpp (hunk 6 above)
Comment (cfjchu) [reply ttnn/cpp/ttnn/tensor/layout/page_config.cpp]: also wondering about what it means to return default constructed tile here?

File: tests/ttnn/unit_tests/operations/eltwise/backward/utility_funcs.py [hunk 22]
```diff
@@ -110,8 +110,12 @@ def compare_results(tt_tensor, golden_tensor, pcc=0.99):
 def compare_pcc(tt_tensor, golden_tensor, pcc=0.99):
     status = True
     for i in range(len(tt_tensor)):
-        tt_out_tensor = tt_tensor[i].cpu().to(ttnn.ROW_MAJOR_LAYOUT).to_torch()
         pt_out_tensor = golden_tensor[i]
+        tt_out_tensor = tt_tensor[i]
+        if tt_out_tensor.dtype != ttnn.bfloat16:
+            tt_out_tensor = ttnn.typecast(tt_out_tensor, ttnn.bfloat16)
```
Comment (cfjchu) [inline tests/ttnn/unit_tests/operations/eltwise/backward/utility_funcs.py]: why is the extra typecast needed? what if the `tt_tensor` was `float32`?

File: ttnn/cpp/ttnn/tensor/tensor_impl.hpp [hunk 23]
```diff
@@ -247,7 +247,7 @@ void validate_sharded_buffer_allocation(
 //                           Data reader, writer, and initializers
 // ======================================================================================
 
-DeviceBuffer allocate_buffer_on_device(Device* device, const ttnn::SimpleShape& shape, const TensorLayout& layout);
+DeviceBuffer allocate_buffer_on_device(Device* device, const TensorSpec& tensor_spec);
```
Comment (cfjchu) [inline ttnn/cpp/ttnn/tensor/tensor_impl.hpp:250]: fyi @omilyutin-tt

**Review by cfjchu — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/slice/device/slice_op.cpp
Comment (jaykru-tt) [inline ttnn/cpp/ttnn/operations/data_movement/slice/device/slice_op.cpp:1]: nice change :)

File: ttnn/cpp/ttnn/operations/data_movement/repeat/repeat.cpp [hunk 24]
```diff
@@ -99,8 +100,7 @@ ttnn::Tensor RepeatOperation::invoke(
 
             auto padded_to_tiled_shape = ttnn::Shape(sliced_logical_shape.view(),
                                                      tiled_output.get_padded_shape().view());
-            tiled_output.set_shape(padded_to_tiled_shape);
-            return tiled_output;
+            return ttnn::reshape(tiled_output, padded_to_tiled_shape);
```
[the remark this answers is not in the document]
Comment (jaykru-tt) [reply ttnn/cpp/ttnn/operations/data_movement/repeat/repeat.cpp:103]: looks fine to me

**Review by jaykru-tt — APPROVED** [review]
TM changes look good to me.

File: tt_metal/tt_stl/reflection.hpp (hunk 17 above)
Comment (ayerofieiev-tt) [reply tt_metal/tt_stl/reflection.hpp:1045]: I see. So you have 2 structures with different meanings, but their hashing otherwise is the same. Or say you have 2 different empty structures. So you must hash the type in. Thank you for the clarification.

File: ttnn/cpp/ttnn/tensor/tensor_spec.hpp (hunk 19 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/tensor_spec.hpp]: I think they should be present in TensorLayout, TensorSpec provides convenience methods and caches some of the results.

File: tt_metal/tt_stl/reflection.hpp (hunk 17 above)
Comment (yan-zaretskiy) [reply tt_metal/tt_stl/reflection.hpp:1045]: Gotcha, I don't have anything against this change. Out of curiosity, I tried this:

```rust
use std::hash::{DefaultHasher, Hash, Hasher};

#[derive(Hash)]
struct RowMajorPageConfig {
    x: u32
}

#[derive(Hash)]
struct TilePageConfig {
    x: u32
}

#[derive(Hash)]
enum Config {
    RowMajorPageConfig(RowMajorPageConfig),
    TilePageConfig(TilePageConfig)
}

fn calculate_hash<T: Hash>(t: &T) -> u64 {
    let mut s = DefaultHasher::new();
    t.hash(&mut s);
    s.finish()
}


pub fn main() {
    println!("{}", calculate_hash(&RowMajorPageConfig{x: 10}));
    println!("{}", calculate_hash(&TilePageConfig{x: 10}));
    println!("{}", calculate_hash(&Config::TilePageConfig(TilePageConfig{x: 10})));
}
```

and I got

```
7396300056459196583
7396300056459196583
17156080100134691638
```

So yeah, Rust hashes an enum with the value differently from the value itself, so we're good!

Specifically in regards to `RowMajorPageConfig`, there are no member variables in that class, so I think what you get is just the hash of the default seed value. So it would be the same for any other empty class. All I'm saying is that this fix might not be enough to correctly handle such situations. It could be better to create a custom hash method.

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp (hunk 18 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp]: I think there was, looking into it more now

File: ttnn/cpp/ttnn/tensor/layout/page_config.cpp (hunk 6 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/layout/page_config.cpp]: This code is no longer here. It was replaced with storing Tile in RowMajor with a comment:
```
 // This is currently needed for compatibility reasons.
 // Each time tile is specified, a warning will be issued. This should be removed soon.
```

File: ttnn/cpp/ttnn/operations/ccl/ccl_common.hpp (hunk 16 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/operations/ccl/ccl_common.hpp]: Yes, thanks

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 3 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:706]: @ntarafdar @jvegaTT What do you think about it? This supposed to be just a view, is there a way to guarantee/check that it is indeed the case?

**Review by yan-zaretskiy — APPROVED** [review]

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp (hunk 18 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp]: The reason for this change is that otherwise `initialize_alignment` -> `create_default_alignment` would be called. Currently it creates a lot of issues across the codebase. I think this would disappear when we migrate all of our code to not use `fromLegacyPaddedShape`

Comment (TT-BrianLiu) [reply ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp]: `initialize_alignment` -> `create_default_alignment` would happen after this change not before right?

File: ttnn/cpp/ttnn/tensor/tensor.hpp [hunk 25]
```diff
@@ -86,21 +83,16 @@ struct Tensor {
...
+    Tensor(Storage storage, TensorSpec tensor_spec);
 
-    // Constructor to initialize unpopulated tensor with workers and storage specified. Use this when creating tensor
+    // Constructors to initialize unpopulated tensor with workers and storage specified. Use this when creating tensor
     // handles in async mode.
-    Tensor(
-        const std::vector<Device *>& workers,
-        uint32_t num_buffers = 0,
-        std::optional<DistributedTensorConfig> distributed_tensor_config = std::nullopt);
+    explicit Tensor(uint32_t num_buffers, std::optional<DistributedTensorConfig> distributed_tensor_config = std::nullopt);
```
Comment (omilyutin-tt) [inline ttnn/cpp/ttnn/tensor/tensor.hpp:94]: For this constructor and the one below - do you want to mention in the docstring that the first one creates a host tensor while the one below is for device tensor?

**Review by pavlejosipovic — APPROVED** [review]

File: ttnn/cpp/ttnn/tensor/tensor.hpp (hunk 25 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/tensor.hpp:94]: Let's leave those two as-is. I expect that both of those constructors will be changed soon
